be/src/exec/pipeline/pipeline_driver_queue.cpp

```diff
@@ -273,186 +174,233 @@ DriverRawPtr SubQuerySharedDriverQueue::take() {
+void WorkGroupDriverQueue::update_statistics(const DriverRawPtr driver) {
     // TODO: reduce the lock scope
-    std::unique_lock<std::mutex> lock(_global_mutex);
+    std::lock_guard<std::mutex> lock(_global_mutex);
 
     int64_t runtime_ns = driver->driver_acct().get_last_time_spent();
-    auto* wg = driver->workgroup();
-    wg->driver_queue()->update_statistics(driver);
-    wg->increment_real_runtime_ns(runtime_ns);
-    workgroup::WorkGroupManager::instance()->increment_cpu_runtime_ns(runtime_ns);
-
-    // For big query check cpu
-    if (wg->big_query_cpu_second_limit()) {
-        // Calculate the cost of the source && sink operator alone
-        int64_t source_operator_last_cpu_time_ns = driver->source_operator()->get_last_growth_cpu_time_ns();
-        int64_t sink_operator_last_cpu_time_ns = driver->sink_operator()->get_last_growth_cpu_time_ns();
-        int64_t accounted_cpu_cost = runtime_ns + source_operator_last_cpu_time_ns + sink_operator_last_cpu_time_ns;
-
-        wg->incr_total_cpu_cost(accounted_cpu_cost);
+    auto* wg_entity = driver->workgroup()->driver_sched_entity();
+
+    // Update bandwidth control information.
+    _update_bandwidth_control_period();
+    if (!wg_entity->is_rt_wg()) {
+        _bandwidth_usage_ns += runtime_ns;
+    }
+
+    // Update sched entity information.
+    bool is_in_queue = _ready_wg_entities.find(wg_entity) != _ready_wg_entities.end();
+    if (is_in_queue) {
+        _ready_wg_entities.erase(wg_entity);
     }
+    DCHECK(_ready_wg_entities.find(wg_entity) == _ready_wg_entities.end());
+    wg_entity->incr_runtime_ns(runtime_ns);
+    if (is_in_queue) {
+        _ready_wg_entities.emplace(wg_entity);
+        _update_min_wg();
+    }
+
+    wg_entity->queue()->update_statistics(driver);
 }
 
-size_t DriverQueueWithWorkGroup::size() const {
+size_t WorkGroupDriverQueue::size() const {
     // TODO: reduce the lock scope
-    std::unique_lock<std::mutex> lock(_global_mutex);
+    std::lock_guard<std::mutex> lock(_global_mutex);
 
     size_t size = 0;
-    for (auto wg : _ready_wgs) {
-        size += wg->driver_queue()->size();
+    for (auto wg_entity : _ready_wg_entities) {
+        size += wg_entity->queue()->size();
     }
 
     return size;
 }
 
+bool WorkGroupDriverQueue::should_yield(const DriverRawPtr driver, int64_t unaccounted_runtime_ns) const {
+    if (_throttled(driver->workgroup()->driver_sched_entity(), unaccounted_runtime_ns)) {
+        return true;
+    }
+
+    // Return true, if the minimum-vruntime workgroup is not current workgroup anymore.
+    auto* wg_entity = driver->workgroup()->driver_sched_entity();
+    return _min_wg_entity.load() != wg_entity &&
+           _min_vruntime_ns.load() < wg_entity->vruntime_ns() + unaccounted_runtime_ns / wg_entity->cpu_limit();
+}
+
+bool WorkGroupDriverQueue::_throttled(const workgroup::WorkGroupDriverSchedEntity* wg_entity,
+                                      int64_t unaccounted_runtime_ns) const {
```

Comment:
what's the meaning of `unaccounted_runtime_ns` ?

Reply:
The runtime not accounted in `wg_entity->runtime()`